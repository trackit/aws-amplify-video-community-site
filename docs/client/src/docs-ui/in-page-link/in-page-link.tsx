import {h, Component, Prop, Watch} from "@stencil/core";
import {headingIsVisible} from "../../utils/heading-is-visible";
import {pageContext} from "../page/page.context";
import {SelectedFilters} from "../page/page.types";
import {getElementTop} from "../../utils/get-element-top";
import {getNavHeight} from "../../utils/get-nav-height";

@Component({tag: "docs-in-page-link", shadow: false})
export class DocsInPageLink {
  /*** the id of the element this link points to */
  @Prop() readonly targetId: string;
  /*** the selected filter state */
  @Prop() readonly selectedFilters: SelectedFilters;

  target?: HTMLElement;

  @Watch("selectedFilters")
  @Watch("targetId")
  componentDidRender() {
    if (this.targetId) {
      const targets = Array.from(
        document.querySelectorAll(`#${this.targetId}`),
      ).filter((e) => headingIsVisible(e)) as HTMLElement[];
      this.target = targets[0] || undefined;
    }
  }

  onClick = (e: Event) => {
    e.preventDefault();
    if (this.target) {
      const top = getElementTop(this.target, getNavHeight("px"));
      if (top !== window.scrollY) {
        window.scrollTo({top});
        history.replaceState(undefined, document.title, `#${this.targetId}`);
      }
    }
  };

  render() {
    return (
      <a onClick={this.onClick}>
        <slot />
      </a>
    );
  }
}

pageContext.injectProps(DocsInPageLink, ["selectedFilters"]);
