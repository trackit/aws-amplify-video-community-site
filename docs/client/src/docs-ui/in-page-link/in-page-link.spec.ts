import {DocsInPageLink} from "./in-page-link";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-in-page-link", () => {
  describe("Component logic", () => {
    let inPageLink: DocsInPageLink;
    beforeEach(() => (inPageLink = new DocsInPageLink()));

    it("should init `targetId` as `undefined`", () =>
      expect(inPageLink.targetId).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsInPageLink],
            html: `<docs-in-page-link />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
