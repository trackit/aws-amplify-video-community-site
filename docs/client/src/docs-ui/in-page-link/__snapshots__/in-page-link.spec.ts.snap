// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-in-page-link Render logic should render 1`] = `
<docs-in-page-link>
  <!---->
  <a></a>
</docs-in-page-link>
`;
